substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool/tx_mem_pool_map.rs

```diff
@@ -0,0 +1,766 @@
+		self.items_by_hashes.contains_key(key)
+	}
+
+	/// Returns an iterator over the values in the map.
+	pub fn values(&self) -> std::collections::hash_map::Values<K, V> {
+		self.items_by_hashes.values()
+	}
+
+	/// Returns the number of elements in the map.
+	pub fn len(&self) -> usize {
+		debug_assert_eq!(self.items_by_hashes.len(), self.items_by_priority.len());
+		self.items_by_hashes.len()
+	}
+
+	/// Removes and returns the first entry in the priority map, if it exists (testing only).
+	#[cfg(test)]
+	pub fn pop_first(&mut self) -> Option<V> {
+		self.items_by_priority.pop_first().map(|(_, v)| v)
+	}
+
+	pub fn with_items<F, R>(&self, f: F) -> R
+	where
+		F: Fn(std::collections::hash_map::Iter<K, V>) -> R,
+	{
+		f(self.items_by_hashes.iter())
+	}
+}
+
+impl<K, A, B, V> IndexedStorage<K, PriorityKey<A, B>, V>
+where
+	K: Ord + std::hash::Hash + Copy,
+	A: Ord,
+	B: Ord,
+	V: Clone + PriorityAndTimestamp<Priority = A, Timestamp = B>,
+	V: std::cmp::PartialEq + std::fmt::Debug,
+{
+	/// Inserts a key-value pair into the map, ordering by priority.
+	pub fn insert(&mut self, key: K, val: V) -> Option<V> {
+		let r = self.items_by_hashes.insert(key, val.clone());
+
+		if let Some(ref removed) = r {
+			let a = self.items_by_priority.remove(&SortKey::new(&key, removed));
+			debug_assert_eq!(r, a);
+		}
+		let a = self.items_by_priority.insert(SortKey::new(&key, &val), val);
+		debug_assert!(a.is_none());
+		r
+	}
+
+	/// Removes a key-value pair from the map based on the key.
+	pub fn remove(&mut self, key: &K) -> Option<V> {
+		let r = self.items_by_hashes.remove(key);
+		let _ = r.as_ref().map(|r| {
+			let k = SortKey::new(key, r);
+			let a = self.items_by_priority.remove(&k);
+			debug_assert_eq!(r.clone(), a.expect("item should be in both maps. qed."));
+		});
+		r
+	}
+
+	/// Allows to mutate item for given key with a closure, if key exists.
+	///
+	/// Intended to mutate priority and timestamp. Changing size is not possible.
+	pub fn update_item<F>(&mut self, key: &K, f: F) -> Option<()>
+	where
+		F: FnOnce(&mut V),
+	{
+		let item = self.items_by_hashes.get_mut(key)?;
+
+		let old_key = SortKey::new(key, item);
+		f(item);
+		let new_key = SortKey::new(key, item);
+
+		if old_key != new_key {
+			self.items_by_priority.remove(&old_key);
+			self.items_by_priority.insert(new_key, item.clone());
+		}
+
+		Some(())
+	}
+}
+
+impl<K, A, B, V> IndexedStorage<K, PriorityKey<A, B>, V>
+where
+	K: Ord + std::hash::Hash + Copy + std::fmt::Debug,
+	A: Ord + std::fmt::Debug,
+	B: Ord + std::fmt::Debug,
+	V: Clone + PriorityAndTimestamp<Priority = A, Timestamp = B> + Size,
+	V: std::cmp::PartialEq + std::fmt::Debug,
+{
+	/// Attempts to insert an item with replacement based on free space and priority.
+	/// Returns the total size in bytes of removed items, and their keys.
+	///
+	/// Insertion always results with other item's removal, the len bound is kept elsewhere
+	///
+	/// (todo which may required some cleanup)
```

Comment:
not super clear what needs to be done in the todo

Reply:
ad3c17e